### PR TITLE
[DOC] Added doxygen comments from spec

### PR DESCRIPTION
**Pull request by outoftardis**
# Description

From #1074: the part with doxygen comments 

- transferred doxygen comments from specs, made small changes in wording (e.g. removed "shall")

RST files will be added separately. The tool is in #1303 and will also be merged separately.

### CODE REVIEW

**Comment by outoftardis** [comment]
@rlnx @michael-smirnov @mchernov-intel  any comments? can we merge this?

**Review by mchernov-intel — APPROVED** [review]
Please, wait for another approve

**Comment by outoftardis** [comment]
@rlnx  @michael-smirnov  any comments from your side?

File: cpp/oneapi/dal/array.hpp [hunk 1]
```diff
@@ -24,6 +24,10 @@
 namespace oneapi::dal {
 namespace v1 {
 
+/// @tparam T The type of the memory block elements within the array.
+///              $T$ can represent any type.
```
Comment (rlnx) [inline cpp/oneapi/dal/array.hpp:28]:
```suggestion
///              :literal:`T` can represent any type.
```

`$...$` is used only for math expressions, `T` here is template parameter

File: cpp/oneapi/dal/array.hpp [hunk 2]
```diff
@@ -24,6 +24,10 @@
...
+///
+/// @pre $T$ cannot be const-qualified.
```
Comment (rlnx) [inline cpp/oneapi/dal/array.hpp:30]:
```suggestion
/// @pre T cannot be const-qualified.
```

File: cpp/oneapi/dal/array.hpp [hunk 3]
```diff
@@ -37,13 +41,28 @@ class array {
     using data_t = T;
 
 public:
+    /// Allocates a new memory block for mutable data, does not initialize it,
+    /// creates a new array instance by passing a pointer to the memory block.
+    /// The array owns the memory block (for details, see :txtref:`data_ownership_requirements`).
+    ///
+    /// @param count The number of elements of type $Data$ to allocate memory for.
```
Comment (rlnx) [inline cpp/oneapi/dal/array.hpp:48]:
```suggestion
    /// @param count The number of elements of type :literal:`Data` to allocate memory for.
```

**Review by rlnx — COMMENTED** [review]
Sometimes `$...$` is used for template or runtime parameters, please replace with :literal:`...`. 

`$...$` should be used only for math expressions.

I've highlighted few cases, but it could be more

**Review by rlnx — APPROVED** [review]

**Pull request merged**